### ISSUE TITLE
Windows SID conversion from string => SID => string gives wrong values

### ISSUE THREAD

**Issue by mharj** [ldaptools/ldaptools#3]
Some reason I can't get  string => SID => string conversion working correctly
As example BUILTIN\ADMINISTRATORS (source https://support.microsoft.com/en-us/kb/163846)
```php
echo fromLdap( toLdap("S-1-5-32-544") );
```
I think this should create same value? (when commented out ldap backslash split/implode part out)

**Comment by ChadSikorra** [comment]
Yeah, it definitely should. I think the `toLdap($sid)` conversion should probably just pass on the string, now that I think of it. For example, the following query where np conversion is done returns just fine:

```
$name = $ldap->buildLdapQuery()
	->select('cn')
    ->where(['objectsid' => 'S-1-5-32-544'])
    ->getLdapQuery()
    ->getSingleScalarResult();
```

And it's not really valid to modify the SID of a LDAP object, so there's no point in a conversion to LDAP. However, converting to a string SID from LDAP binary data still doesn't work for Builtin LDAP objects. I probably made too many assumptions regarding the structure of the data. I'll have to see how to go about fixing that.

**Comment by mharj** [comment]
Yeah, I think you can lookup AD directly with SID string.
Basically I'm building cache system for SID:s (which uses binary version) so I need to load less from AD.
So I tried to pre fill cache with toLdap conversion and some BUILTIN SID:s ... and in conversion I noticed that I did get some strange values :)

**Comment by ChadSikorra** [comment]
Actually, I just realized that calling `echo $converter->fromLdap($converter->toLdap("S-1-5-32-544"));` will never work. The reason is because the value from `toLdap($sid)` needs to be a string SID or hex escaped string. Those are the only valid SID values going to LDAP via a query, which is really the only function for that converter since the SID is a system value and cannot be changed directly.

To simulate the "raw" data (like how it comes from LDAP and is passed to the converter initially) you could do something like: `$converter->fromLdap(pack('H*', str_replace('\\','',$converter->toLdap('S-1-5-32-544'))));`. But it sounds to me like you may have already done this to some extent (replacing the slashes that is)?

However, there is still an issue with the conversion of a builtin SID string to its hex counterpart like you'd see in AD. The issue seems to be with the identifier authority value immediately following the revision number. I'll be taking a closer look at that soon hopefully. But I'm tempted to just scrap the whole conversion process if AD accepts a simple SID string for query values.

**Comment by ChadSikorra** [comment]
So apparently I ignored the subauthority count when I created the converter...whoops. Hah. Anyways, this should now be fixed by https://github.com/ldaptools/ldaptools/commit/84bed81fef58a48e8717202722e769f1f31dc6c7. It was really just dumb luck that it worked for most SIDs in the first place. Interesting reading for SID structure: https://technet.microsoft.com/en-us/library/cc962011.aspx

Anyway, that should fix your problem. Hopefully I'll be tagging a new version soon since I just added the LDIF parser recently.
